lib/min_heap.rb

```diff
@@ -44,28 +49,45 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(1)
+  # Space complexity: O(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    if @store.length == 0
+      return true
+    else
+      return false
+    end
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(log n)
+  # Space complexity: O(1)
   def heap_up(index)
-    
+    parent_i = (index - 1) / 2
+    return if index == 0 || @store[parent_i].key < @store[index].key
+
+    swap(parent_i, index)
+    heap_up(parent_i)
   end
 
   # This helper method takes an index and 
   #  moves it up the heap if it's smaller
   #  than it's parent node.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    left_child = 2 * index + 1
+    right_child = 2 * index + 2
+    return if @store[left_child].nil? || @store[right_child].nil?
```

Comment:
What if `@store[left_child]` is not nil but the right child is?  Could you miss a needed swap?